File: sdks/java/src/main/java/com/hypersim/security/SecurityManager.java
package com.hypersim.security;

import java.util.*;
import java.util.concurrent.*;
import java.time.Duration;
import java.time.Instant;
import java.security.SecureRandom;
import javax.crypto.Cipher;
import javax.crypto.spec.SecretKeySpec;
import javax.crypto.spec.IvParameterSpec;

/**
 * Comprehensive Security Manager for HyperSim Java SDK
 * Orchestrates all security features including API key rotation,
 * multi-signature support, request signing, and OWASP compliance.
 */
public class SecurityManager {
    private final SecurityConfig config;
    private final APIKeyManager apiKeyManager;
    private final MultiSignatureManager multiSigManager;
    private final RequestSigner requestSigner;
    private final RateLimiter rateLimiter;
    private final SecureStorage secureStorage;
    private final SecurityAuditor securityAuditor;
    private final OWASPValidator owaspValidator;
    private final InputSanitizer inputSanitizer;
    private final SecurityMetrics metrics;
    private final ScheduledExecutorService scheduler;
    private volatile boolean initialized = false;
    private final Object lock = new Object();
    
    /**
     * Security configuration class
     */
    public static class SecurityConfig {
        private boolean apiKeyRotation = true;
        private Duration keyRotationInterval = Duration.ofHours(24);
        private boolean multiSigEnabled = false;
        private boolean requestSigning = true;
        private RateLimitConfig rateLimit = new RateLimitConfig();
        private boolean auditLogging = true;
        private boolean owaspCompliance = true;
        private List<String> certificatePins = new ArrayList<>();
        private String inputValidation = "strict";
        private boolean debug = false;
        
        // Getters and setters
        public boolean isApiKeyRotation() { return apiKeyRotation; }
        public void setApiKeyRotation(boolean apiKeyRotation) { this.apiKeyRotation = apiKeyRotation; }
        
        public Duration getKeyRotationInterval() { return keyRotationInterval; }
        public void setKeyRotationInterval(Duration keyRotationInterval) { this.keyRotationInterval = keyRotationInterval; }
        
        public boolean isMultiSigEnabled() { return multiSigEnabled; }
        public void setMultiSigEnabled(boolean multiSigEnabled) { this.multiSigEnabled = multiSigEnabled; }
        
        public boolean isRequestSigning() { return requestSigning; }
        public void setRequestSigning(boolean requestSigning) { this.requestSigning = requestSigning; }
        
        public RateLimitConfig getRateLimit() { return rateLimit; }
        public void setRateLimit(RateLimitConfig rateLimit) { this.rateLimit = rateLimit; }
        
        public boolean isAuditLogging() { return auditLogging; }
        public void setAuditLogging(boolean auditLogging) { this.auditLogging = auditLogging; }
        
        public boolean isOwaspCompliance() { return owaspCompliance; }
        public void setOwaspCompliance(boolean owaspCompliance) { this.owaspCompliance = owaspCompliance; }
        
        public List<String> getCertificatePins() { return certificatePins; }
        public void setCertificatePins(List<String> certificatePins) { this.certificatePins = certificatePins; }
        
        public String getInputValidation() { return inputValidation; }
        public void setInputValidation(String inputValidation) { this.inputValidation = inputValidation; }
        
        public boolean isDebug() { return debug; }
        public void setDebug(boolean debug) { this.debug = debug; }
    }
    
    /**
     * Rate limit configuration
     */
    public static class RateLimitConfig {
        private int requestsPerWindow = 100;
        private Duration windowDuration = Duration.ofMinutes(1);
        private int maxQueueSize = 50;
        private boolean ddosProtection = true;
        private int ddosThreshold = 1000;
        
        // Getters and setters
        public int getRequestsPerWindow() { return requestsPerWindow; }
        public void setRequestsPerWindow(int requestsPerWindow) { this.requestsPerWindow = requestsPerWindow; }
        
        public Duration getWindowDuration() { return windowDuration; }
        public void setWindowDuration(Duration windowDuration) { this.windowDuration = windowDuration; }
        
        public int getMaxQueueSize() { return maxQueueSize; }
        public void setMaxQueueSize(int maxQueueSize) { this.maxQueueSize = maxQueueSize; }
        
        public boolean isDdosProtection() { return ddosProtection; }
        public void setDdosProtection(boolean ddosProtection) { this.ddosProtection = ddosProtection; }
        
        public int getDdosThreshold() { return ddosThreshold; }
        public void setDdosThreshold(int ddosThreshold) { this.ddosThreshold = ddosThreshold; }
    }
    
    /**
     * Security metrics tracking
     */
    public static class SecurityMetrics {
        private volatile long totalRequests = 0;
        private volatile long blockedRequests = 0;
        private volatile long rateLimitViolations = 0;
        private volatile long failedSignatures = 0;
        private volatile long owaspViolations = 0;
        private volatile long keysRotated = 0;
        
        // Getters
        public long getTotalRequests() { return totalRequests; }
        public long getBlockedRequests() { return blockedRequests; }
        public long getRateLimitViolations() { return rateLimitViolations; }
        public long getFailedSignatures() { return failedSignatures; }
        public long getOwaspViolations() { return owaspViolations; }
        public long getKeysRotated() { return keysRotated; }
        
        // Increment methods
        public void incrementTotalRequests() { totalRequests++; }
        public void incrementBlockedRequests() { blockedRequests++; }
        public void incrementRateLimitViolations() { rateLimitViolations++; }
        public void incrementFailedSignatures() { failedSignatures++; }
        public void incrementOwaspViolations() { owaspViolations++; }
        public void incrementKeysRotated() { keysRotated++; }
    }
    
    /**
     * Constructor
     */
    public SecurityManager(SecurityConfig config) {
        this.config = config != null ? config : new SecurityConfig();
        this.metrics = new SecurityMetrics();
        this.scheduler = Executors.newScheduledThreadPool(2);
        
        // Initialize components
        this.secureStorage = new SecureStorage();
        this.apiKeyManager = new APIKeyManager(
            secureStorage, 
            this.config.isApiKeyRotation(), 
            this.config.getKeyRotationInterval()
        );
        this.multiSigManager = new MultiSignatureManager(secureStorage);
        this.requestSigner = new RequestSigner(secureStorage);
        this.rateLimiter = new RateLimiter(this.config.getRateLimit());
        this.securityAuditor = new SecurityAuditor(this.config.isAuditLogging());
        this.owaspValidator = new OWASPValidator(this.config.isOwaspCompliance());
        this.inputSanitizer = new InputSanitizer(this.config.getInputValidation());
    }
    
    /**
     * Initialize all security components
     */
    public CompletableFuture<Void> initialize() {
        return CompletableFuture.runAsync(() -> {
            synchronized (lock) {
                if (initialized) {
                    return;
                }
                
                try {
                    // Initialize all components
                    secureStorage.initialize();
                    apiKeyManager.initialize();
                    multiSigManager.initialize();
                    requestSigner.initialize();
                    rateLimiter.initialize();
                    securityAuditor.initialize();
                    owaspValidator.initialize();
                    inputSanitizer.initialize();
                    
                    // Start background tasks
                    if (config.isApiKeyRotation()) {
                        scheduleKeyRotation();
                    }
                    
                    scheduleSecurityMonitoring();
                    
                    initialized = true;
                    
                    if (config.isDebug()) {
                        System.out.println("[HyperSim Security] Security manager initialized");
                    }
                    
                } catch (Exception e) {
                    throw new RuntimeException("Failed to initialize security manager", e);
                }
            }
        });
    }
    
    /**
     * Process and secure an outgoing request
     */
    public CompletableFuture<Map<String, Object>> secureRequest(Map<String, Object> request) {
        return CompletableFuture.supplyAsync(() -> {
            metrics.incrementTotalRequests();
            
            try {
                // Rate limiting check
                if (!rateLimiter.isAllowed(request)) {
                    metrics.incrementRateLimitViolations();
                    metrics.incrementBlockedRequests();
                    throw new SecurityException("Rate limit exceeded");
                }
                
                // Input sanitization
                Map<String, Object> sanitized = inputSanitizer.sanitize(request);
                
                // OWASP compliance check
                if (config.isOwaspCompliance()) {
                    List<String> violations = owaspValidator.validate(sanitized);
                    if (!violations.isEmpty()) {
                        metrics.incrementOwaspViolations();
                        throw new SecurityException("OWASP violations detected: " + violations);
                    }
                }
                
                // Request signing
                Map<String, Object> finalRequest = sanitized;
                if (config.isRequestSigning()) {
                    finalRequest = requestSigner.signRequest(sanitized);
                }
                
                // Add API key
                APIKeyData apiKey = apiKeyManager.getCurrentKey();
                Map<String, String> headers = (Map<String, String>) finalRequest.getOrDefault("headers", new HashMap<>());
                headers.put("Authorization", "Bearer " + apiKey.getPrimary());
                headers.put("X-API-Key-Rotation", String.valueOf(apiKey.getRotationCount()));
                finalRequest.put("headers", headers);
                
                return finalRequest;
                
            } catch (Exception e) {
                metrics.incrementBlockedRequests();
                throw new RuntimeException("Request security processing failed", e);
            }
        });
    }
    
    /**
     * Get current security metrics
     */
    public SecurityMetrics getMetrics() {
        return metrics;
    }
    
    /**
     * Manually rotate API keys
     */
    public CompletableFuture<Void> rotateApiKeys() {
        return CompletableFuture.runAsync(() -> {
            try {
                apiKeyManager.rotateKeys();
                metrics.incrementKeysRotated();
            } catch (Exception e) {
                throw new RuntimeException("API key rotation failed", e);
            }
        });
    }
    
    /**
     * Shutdown security manager
     */
    public CompletableFuture<Void> shutdown() {
        return CompletableFuture.runAsync(() -> {
            synchronized (lock) {
                if (!initialized) {
                    return;
                }
                
                try {
                    scheduler.shutdown();
                    if (!scheduler.awaitTermination(5, TimeUnit.SECONDS)) {
                        scheduler.shutdownNow();
                    }
                    
                    // Shutdown all components
                    securityAuditor.flush();
                    
                    initialized = false;
                    
                    if (config.isDebug()) {
                        System.out.println("[HyperSim Security] Security manager shut down");
                    }
                    
                } catch (Exception e) {
                    throw new RuntimeException("Failed to shutdown security manager", e);
                }
            }
        });
    }
    
    private void scheduleKeyRotation() {
        scheduler.scheduleAtFixedRate(() -> {
            try {
                apiKeyManager.rotateKeys();
                metrics.incrementKeysRotated();
            } catch (Exception e) {
                if (config.isDebug()) {
                    System.err.println("[HyperSim Security] Key rotation failed: " + e.getMessage());
                }
            }
        }, config.getKeyRotationInterval().toMinutes(), config.getKeyRotationInterval().toMinutes(), TimeUnit.MINUTES);
    }
    
    private void scheduleSecurityMonitoring() {
        scheduler.scheduleAtFixedRate(() -> {
            try {
                checkSuspiciousActivities();
            } catch (Exception e) {
                if (config.isDebug()) {
                    System.err.println("[HyperSim Security] Security monitoring error: " + e.getMessage());
                }
            }
        }, 30, 30, TimeUnit.SECONDS);
    }
    
    private void checkSuspiciousActivities() {
        // Implementation for detecting suspicious activities
        // This would check rate limit violations, failed authentications, etc.
        
        if (metrics.getRateLimitViolations() > config.getRateLimit().getDdosThreshold()) {
            securityAuditor.logSecurityEvent("Potential DDoS attack detected", "HIGH");
        }
    }
    
    // Placeholder classes for other security components
    private static class APIKeyManager {
        public APIKeyManager(SecureStorage storage, boolean rotationEnabled, Duration interval) {}
        public void initialize() {}
        public APIKeyData getCurrentKey() { return new APIKeyData(); }
        public void rotateKeys() {}
    }
    
    private static class APIKeyData {
        public String getPrimary() { return "mock-api-key"; }
        public int getRotationCount() { return 1; }
    }
    
    private static class MultiSignatureManager {
        public MultiSignatureManager(SecureStorage storage) {}
        public void initialize() {}
    }
    
    private static class RequestSigner {
        public RequestSigner(SecureStorage storage) {}
        public void initialize() {}
        public Map<String, Object> signRequest(Map<String, Object> request) { return request; }
    }
    
    private static class RateLimiter {
        public RateLimiter(RateLimitConfig config) {}
        public void initialize() {}
        public boolean isAllowed(Map<String, Object> request) { return true; }
    }
    
    private static class SecureStorage {
        public void initialize() {}
    }
    
    private static class SecurityAuditor {
        public SecurityAuditor(boolean enabled) {}
        public void initialize() {}
        public void flush() {}
        public void logSecurityEvent(String message, String severity) {}
    }
    
    private static class OWASPValidator {
        public OWASPValidator(boolean enabled) {}
        public void initialize() {}
        public List<String> validate(Map<String, Object> data) { return new ArrayList<>(); }
    }
    
    private static class InputSanitizer {
        public InputSanitizer(String level) {}
        public void initialize() {}
        public Map<String, Object> sanitize(Map<String, Object> data) { return data; }
    }
}

File: sdks/go/security/security_manager.go
// Package security provides comprehensive security features for HyperSim Go SDK
package security

import (
	"context"
	"crypto/rand"
	"encoding/hex"
	"fmt"
	"sync"
	"time"
)

// SecurityManager orchestrates all security features
type SecurityManager struct {
	config          *SecurityConfig
	apiKeyManager   *APIKeyManager
	multiSig        *MultiSignature
	requestSigner   *RequestSigner
	rateLimiter     *RateLimiter
	secureStorage   *SecureStorage
	securityAuditor *SecurityAuditor
	owaspValidator  *OWASPValidator
	inputSanitizer  *InputSanitizer
	metrics         *SecurityMetrics
	mu              sync.RWMutex
	initialized     bool
	shutdownCh      chan struct{}
}

// SecurityConfig holds security configuration
type SecurityConfig struct {
	APIKeyRotation     bool
	KeyRotationInterval time.Duration
	MultiSigEnabled    bool
	RequestSigning     bool
	RateLimit          *RateLimitConfig
	AuditLogging       bool
	OWASPCompliance    bool
	CertificatePins    []string
	InputValidation    string // "strict", "moderate", "basic"
	Debug              bool
}

// SecurityMetrics tracks security events
type SecurityMetrics struct {
	TotalRequests       int64 `json:"total_requests"`
	BlockedRequests     int64 `json:"blocked_requests"`
	RateLimitViolations int64 `json:"rate_limit_violations"`
	FailedSignatures    int64 `json:"failed_signatures"`
	OWASPViolations     int64 `json:"owasp_violations"`
	KeysRotated         int64 `json:"keys_rotated"`
}

// NewSecurityManager creates a new security manager
func NewSecurityManager(config *SecurityConfig) *SecurityManager {
	if config == nil {
		config = DefaultSecurityConfig()
	}
	
	sm := &SecurityManager{
		config:     config,
		metrics:    &SecurityMetrics{},
		shutdownCh: make(chan struct{}),
	}
	
	// Initialize components
	sm.secureStorage = NewSecureStorage()
	sm.apiKeyManager = NewAPIKeyManager(sm.secureStorage, config.APIKeyRotation, config.KeyRotationInterval)
	sm.multiSig = NewMultiSignature(sm.secureStorage)
	sm.requestSigner = NewRequestSigner(sm.secureStorage)
	sm.rateLimiter = NewRateLimiter(config.RateLimit)
	sm.securityAuditor = NewSecurityAuditor(config.AuditLogging)
	sm.owaspValidator = NewOWASPValidator(config.OWASPCompliance)
	sm.inputSanitizer = NewInputSanitizer(config.InputValidation)
	
	return sm
}

// Initialize starts all security components
func (sm *SecurityManager) Initialize(ctx context.Context) error {
	sm.mu.Lock()
	defer sm.mu.Unlock()
	
	if sm.initialized {
		return nil
	}
	
	// Initialize all components
	if err := sm.secureStorage.Initialize(ctx); err != nil {
		return fmt.Errorf("failed to initialize secure storage: %w", err)
	}
	
	if err := sm.apiKeyManager.Initialize(ctx); err != nil {
		return fmt.Errorf("failed to initialize API key manager: %w", err)
	}
	
	// Start background tasks
	if sm.config.APIKeyRotation {
		go sm.autoRotateKeys(ctx)
	}
	
	go sm.securityMonitor(ctx)
	
	sm.initialized = true
	return nil
}

// SecureRequest processes and secures an outgoing request
func (sm *SecurityManager) SecureRequest(ctx context.Context, request map[string]interface{}) (map[string]interface{}, error) {
	sm.metrics.TotalRequests++
	
	// Rate limiting check
	if !sm.rateLimiter.IsAllowed(request) {
		sm.metrics.RateLimitViolations++
		sm.metrics.BlockedRequests++
		return nil, fmt.Errorf("rate limit exceeded")
	}
	
	// Input sanitization
	sanitized, err := sm.inputSanitizer.Sanitize(request)
	if err != nil {
		return nil, fmt.Errorf("input sanitization failed: %w", err)
	}
	
	// OWASP compliance check
	if sm.config.OWASPCompliance {
		violations := sm.owaspValidator.Validate(sanitized)
		if len(violations) > 0 {
			sm.metrics.OWASPViolations += int64(len(violations))
			return nil, fmt.Errorf("OWASP violations detected: %v", violations)
		}
	}
	
	// Request signing
	finalRequest := sanitized
	if sm.config.RequestSigning {
		finalRequest, err = sm.requestSigner.SignRequest(sanitized)
		if err != nil {
			return nil, fmt.Errorf("request signing failed: %w", err)
		}
	}
	
	// Add API key
	apiKey, err := sm.apiKeyManager.GetCurrentKey()
	if err != nil {
		return nil, fmt.Errorf("failed to get API key: %w", err)
	}
	
	headers, ok := finalRequest["headers"].(map[string]string)
	if !ok {
		headers = make(map[string]string)
	}
	headers["Authorization"] = fmt.Sprintf("Bearer %s", apiKey.Primary)
	headers["X-API-Key-Rotation"] = fmt.Sprintf("%d", apiKey.RotationCount)
	finalRequest["headers"] = headers
	
	return finalRequest, nil
}

// GetMetrics returns current security metrics
func (sm *SecurityManager) GetMetrics() *SecurityMetrics {
	sm.mu.RLock()
	defer sm.mu.RUnlock()
	return sm.metrics
}

// Shutdown gracefully stops all security components
func (sm *SecurityManager) Shutdown(ctx context.Context) error {
	sm.mu.Lock()
	defer sm.mu.Unlock()
	
	if !sm.initialized {
		return nil
	}
	
	close(sm.shutdownCh)
	
	// Shutdown all components
	if err := sm.securityAuditor.Flush(); err != nil {
		return fmt.Errorf("failed to flush audit logs: %w", err)
	}
	
	sm.initialized = false
	return nil
}

// autoRotateKeys performs automatic key rotation
func (sm *SecurityManager) autoRotateKeys(ctx context.Context) {
	ticker := time.NewTicker(sm.config.KeyRotationInterval)
	defer ticker.Stop()
	
	for {
		select {
		case <-ctx.Done():
			return
		case <-sm.shutdownCh:
			return
		case <-ticker.C:
			if err := sm.apiKeyManager.RotateKeys(); err != nil {
				// Log error but continue
				continue
			}
			sm.metrics.KeysRotated++
		}
	}
}

// securityMonitor monitors for security threats
func (sm *SecurityManager) securityMonitor(ctx context.Context) {
	ticker := time.NewTicker(30 * time.Second)
	defer ticker.Stop()
	
	for {
		select {
		case <-ctx.Done():
			return
		case <-sm.shutdownCh:
			return
		case <-ticker.C:
			// Check for suspicious activities
			sm.checkSuspiciousActivities()
		}
	}
}

func (sm *SecurityManager) checkSuspiciousActivities() {
	// Implementation for detecting suspicious activities
	// This would check rate limit violations, failed authentications, etc.
}

// DefaultSecurityConfig returns default security configuration
func DefaultSecurityConfig() *SecurityConfig {
	return &SecurityConfig{
		APIKeyRotation:      true,
		KeyRotationInterval: 24 * time.Hour,
		MultiSigEnabled:     false,
		RequestSigning:      true,
		RateLimit: &RateLimitConfig{
			RequestsPerWindow: 100,
			WindowDuration:    time.Minute,
			MaxQueueSize:      50,
		},
		AuditLogging:    true,
		OWASPCompliance: true,
		InputValidation: "strict",
		Debug:           false,
	}
}